app/server/lib/GristSocketServer.ts

```diff
@@ -30,6 +30,23 @@ export class GristSocketServerEIO extends GristSocketServer {
       allowUpgrades: false,
       transports: ['polling', 'websocket'],
       maxHttpBufferSize: MAX_PAYLOAD,
+      cors: {
+        // This will cause engine.io to reflect any client-provided Origin into
+        // the Access-Control-Allow-Origin header, essentially disabling the
+        // protection offered by the Same-Origin Policy. This sounds insecure
+        // but is actually the security model of native WebSockets (they are
+        // not covered by SOP; any webpage can open a WebSocket connecting to
+        // any other domain, including the target domain's cookies; it is up to
+        // the receiving server to check the request's Origin header). Since
+        // this CORS behavior will only apply to polling requests, which end up
+        // subjected downstream to the same Origin checks as those necessarily
+        // applied to WebSocket requests, it is safe to let any client attempt
+        // a connection here.
```

Comment:
Ooh, good thing you looked into this then. Indeed, after a bit of testing on my side I can confirm it looks like `SameSite` is what is really keeping cross-origin WebSockets from being a vulnerability (WebSocket requests do include what cookies they can by default, without having to add e.g. `credentials: "include"` as is required with `fetch`).

It looks like the underlying assumption that "Opening up CORS for regular requests that are being handled by the same code that treats WebSocket requests (however insecurely) does not make things worse" still holds though.

But I agree that checking `Origin` would make for a more secure implementation all around. I would not mind looking into including the extra check, however I would require a few pointers to find the proper methods to call for verifying the origin.

The commit you referenced modifies the `FlexServer.trustOriginHandler` method but I cannot call this when validating a WebSocket request, since it is not in the `(req, res, next)` context of an Express middleware. The underlying `requestUtils.trustOrigin` is still bound to a `Request`/`Response` interface and can add response headers which again do not make sense in a WebSocket response (by the way, the fact that on [this line](https://github.com/gristlabs/grist-core/blob/4654edca3dc18376828906fe570e9ff4c69ba8ab/app/server/lib/requestUtils.ts#L89) `req.hostname` is checked rather than `origin` is puzzling).

It looks like the core of what `trustOrigin` does is in this line:

```ts
if (!allowHost(req, new URL(origin)) && !isEnvironmentAllowedHost(new URL(origin)))
```

I could add something like this early in `Comm._onWebSocketConnection` (where I found out WebSocket requests whose `Host` header does not match an allowed value are rejected, as a result of calling `hosts.addOrgInfo`).

But looking at `allowHost`, I see a cast of `req` to `RequestWithOrg` (when the raw WebSocket request clearly has no organization info, at least when it first arrives)  and this gets into a part of the code I have no familiarity with yet. I wouldn't want to break a legitimate scenario, so I'm afraid some hand-holding will be necessary here.

If you can show me how and where you would add this Origin check to WebSocket requests in the current codebase, I think I would easily port this to my rearranged version of the code in this PR. Come to think of it, maybe it would be more prudent to add this check in a separate (prerequisite to this one) PR?

Reply:
> ...require a few pointers to find the proper methods to call for verifying the origin.

What you suggested is exactly what I had in mind. `addOrgInfo` is what turns a `req` into `RequestWithOrg`, so if you add the `allowHost` check afterwards, it should almost make sense. "Almost" because you would need to adjust some methods to accept either `express.Request` or the underlying `http.IncomingMessage`. I think that's always possible by getting headers using `req.headers[lowercase_name]`.

If `allowHost` returns false, we could then either fail when credentials are present (since we don't expect them to be), like `trustOriginHandler` does, or perhaps just strip the credentials and keep going.

> the fact that on this line req.hostname is checked rather than origin is puzzling

Doh, it's puzzling to me too. It looks like a mistake. It seems it was added for tests, and in practice normally has no effect (because GRIST_HOST is either unset or set to something like `0.0.0.0`), but if it did have an effect, it also looks like the wrong one to me. I'd try removing it and seeing if any tests fail. If you can try that, that would be great; if not, let me know, I'll make a note so it doesn't fall through the cracks. Thank you for noticing!

Reply:
>> the fact that on this line req.hostname is checked rather than origin is puzzling
>
> Doh, it's puzzling to me too. It looks like a mistake. It seems it was added for tests, and in practice normally has no effect (because GRIST_HOST is either unset or set to something like 0.0.0.0), but if it did have an effect, it also looks like the wrong one to me. I'd try removing it and seeing if any tests fail. If you can try that, that would be great; if not, let me know, I'll make a note so it doesn't fall through the cracks. Thank you for noticing!

I can confirm that removing this line has no effect on tests. I ran the whole test suite on GitHub Actions, even though it seems only the `DocApi` test suite would catch a regression in this code based on the fact that it seems to be the only suite where the `Origin` header is varied.

So I went ahead and removed that line. Now looking into getting websocket request origins checked as you described. This will involve adding some `Comm` tests I believe.